src/compiler/utilities.ts

```diff
@@ -1628,21 +1630,48 @@ namespace ts {
         "\u0085": "\\u0085"  // nextLine
     };
 
+    let singleQuoteEscapedCharsRegExp = /[\\\'\u0000-\u001f\t\v\f\b\r\n\u2028\u2029\u0085]/g;
+    let singleQuoteEscapedCharsMap: Map<string> = {
+        "\0": "\\0",
+        "\t": "\\t",
+        "\v": "\\v",
+        "\f": "\\f",
+        "\b": "\\b",
+        "\r": "\\r",
+        "\n": "\\n",
+        "\\": "\\\\",
+        "\'": "\\\'",
+        "\u2028": "\\u2028", // lineSeparator
+        "\u2029": "\\u2029", // paragraphSeparator
+        "\u0085": "\\u0085"  // nextLine
+    };
+
     /**
      * Based heavily on the abstract 'Quote'/'QuoteJSONString' operation from ECMA-262 (24.3.2.2),
      * but augmented for a few select characters (e.g. lineSeparator, paragraphSeparator, nextLine)
      * Note that this doesn't actually wrap the input in double quotes.
      */
     export function escapeString(s: string): string {
-        s = escapedCharsRegExp.test(s) ? s.replace(escapedCharsRegExp, getReplacement) : s;
+        return escapeStringByQuote(s, "\"");
+    }
+
+    export function escapeStringByQuote(s: string, quotemark: string): string {

```

Comment:
The reason it's not an enum (yet) is because at use-site it is called with the result of a regex match in the form `('|")`. Naturally, it could have a lookup added in. I don't think that either string literal types or enums make the pattern less cumbersome, though.